Feature/apply state hoisting

atimadanoglu opened the pull request.

atimadanoglu committed:
Created NoteUIMapper and changed NoteMapper's class name as NoteEntityMapper

atimadanoglu committed:
Implemented GetNotesUseCase

atimadanoglu committed:
Implemented GetNoteByIdUseCase

atimadanoglu committed:
Created a new model, changed a file's location

-Created AddNoteRequest model
-Changed location of Note model
-Updated the id property of NoteEntity as nullable

atimadanoglu committed:
Implemented addNoteUseCase

atimadanoglu committed:
Added hilt and viewModel dependencies for compose

atimadanoglu committed:
Implemented MainScreenViewModel

-Added a viewModel parameter to MainScreen composable
-Updated the getNotes query to get the list in descending order

atimadanoglu committed:
Implemented AddNoteViewModel

atimadanoglu committed:
Implemented GetOrUpdateNoteViewModel

atimadanoglu committed:
Updated test cases for notesRepository

atimadanoglu committed:
Implemented test cases for useCases

atimadanoglu committed:
Implemented addNoteScreen and mainScreen navigation

atimadanoglu committed:
Changed NoteUIAndRequestMapper's name as NoteUIMapper

atimadanoglu committed:
Applied state-hoisting on NotesListScreen

-Changed MainScreen's name as NotesListScreen
-Changed MainScreenViewModel's name as NotesListViewModel

atimadanoglu committed:
Applied state-hoisting on AddNoteScreen

-Created a new mapper called AddNoteRequestMapper

atimadanoglu committed:
Changed ui states' location and their name

-Created interfaces called NotesListUiState and AddNoteUiState

atimadanoglu committed:
Implemented searchNotes feature

-Added new property into NotesListUiState

mertceyhan commented on app/src/main/java/com/atakanmadanoglu/notesapplication/data/local/NotesDao.kt:
```diff
@@ -10,7 +10,7 @@ import kotlinx.coroutines.flow.Flow
 interface NotesDao {
 
-    @Query("SELECT * FROM note")
+    @Query("SELECT * FROM note ORDER BY id DESC")
```
`createdAt`

mertceyhan commented on app/src/main/java/com/atakanmadanoglu/notesapplication/data/local/NotesDao.kt:
```diff
@@ -10,7 +10,7 @@ import kotlinx.coroutines.flow.Flow
 
-    @Query("SELECT * FROM note")
+    @Query("SELECT * FROM note ORDER BY id DESC")
     fun getNotes(): Flow<List<NoteEntity>>
```
getNotesByCreatedAt()

mertceyhan commented on app/src/main/java/com/atakanmadanoglu/notesapplication/data/mapper/NoteEntityMapper.kt:
```diff
@@ -1,16 +1,16 @@
 import javax.inject.Inject
 
-class NoteMapper @Inject constructor() {
+class NoteEntityMapper @Inject constructor() {
```
Move the mapper to domain layer

mertceyhan commented on app/src/main/java/com/atakanmadanoglu/notesapplication/data/model/NoteEntity.kt:
```diff
@@ -7,7 +7,7 @@ import androidx.room.PrimaryKey
 data class NoteEntity(
     @PrimaryKey(autoGenerate = true)
-    val id: Int,
+    val id: Int? = null,
```
`val id: Int = 0`

mertceyhan commented on app/src/main/java/com/atakanmadanoglu/notesapplication/data/repository/NotesRepositoryImp.kt:
```diff
@@ -1,15 +1,16 @@
 import kotlinx.coroutines.flow.map
 import javax.inject.Inject
 
 class NotesRepositoryImp @Inject constructor(
```
Repo gets and returns just an `Entity`

mertceyhan commented on app/src/main/java/com/atakanmadanoglu/notesapplication/domain/model/NoteUI.kt:
```diff
@@ -0,0 +1,19 @@
+import java.util.*
+
+data class NoteUI(
+    val id: Int? = null,
```
`val id: Int
`

mertceyhan commented on app/src/main/java/com/atakanmadanoglu/notesapplication/domain/model/NoteUI.kt:
```diff
@@ -0,0 +1,19 @@
+    val id: Int? = null,
+    val title: String,
+    val description: String,
+    val createdAt: Long
```
`val createdAt: String`

mertceyhan commented on app/src/main/java/com/atakanmadanoglu/notesapplication/presentation/add_note/AddNoteViewModel.kt:
```diff
@@ -0,0 +1,75 @@
+    private val _addNoteUiState = MutableStateFlow(MutableAddNoteUiState())
+    val addNoteUiState: StateFlow<AddNoteUiState> = _addNoteUiState.asStateFlow()
+
+    fun addNote(
```
```

    fun addNote(
        inputTitle: String,
        inputDescription: String
    ) = viewModelScope.launch {
        with(_addNoteUiState.value) {
            title = inputTitle
            description = inputDescription
            addNoteUseCase.invoke(
                inputTitle,
                inputDescription
            )
        }
    }
```

mertceyhan commented on app/src/main/java/com/atakanmadanoglu/notesapplication/presentation/get_or_update_note/GetOrUpdateNoteViewModel.kt:
```diff
@@ -0,0 +1,26 @@
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+class GetOrUpdateNoteViewModel @Inject constructor(
```
`EditNoteViewModel`

mertceyhan commented on app/src/main/java/com/atakanmadanoglu/notesapplication/presentation/notes_list/NotesListScreen.kt:
```diff
@@ -12,39 +12,44 @@ import androidx.compose.ui.draw.clip
 @Composable
-fun MainScreen() {
+fun NotesListScreen(
+    navigateToAddNoteScreen: () -> Unit,
```
`addNoteButtonClicked: `

atimadanoglu committed:
Deleted AddNoteRequest and its mapper

-Updated affected parts because of these deletions

atimadanoglu committed:
Refactored the model name in domain layer

-Edited affected parts because of this edit

atimadanoglu committed:
Moved the date formatting operation to domain layer

-Edited noteEntity's id property
-Removed date properties of noteUI
-Changed createdAt property's type of noteUI as String

atimadanoglu committed:
Changed getNotes function's name in dao, repo and useCase

-Changed order_by property in dao
-Changed some names for readability

atimadanoglu committed:
Moved useCases to usecases package under domain package
